### PR TITLE
Training evaluation

### PR DESCRIPTION
**Pull request by luotigerlsx**
1. remove dependency on tensorflow
2. add additional argument parsing for randomforest hyperparameter tuning
3. save the trained model and evluation result

### CODE REVIEW

File: sklearn/sklearn-template/template/setup.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
     'pandas-gbq>=0.8.0',
     'cloudml-hypertune',
     'scikit-plot',
+    "urllib"
```
Comment (cfezequiel) [inline sklearn/sklearn-template/template/setup.py]: Recommend changing to `'urllib',`

File: sklearn/sklearn-template/template/trainer/metadata.py [hunk 2]
```diff
@@ -19,17 +19,20 @@
 
 
 NUMERIC_FEATURES = [
```
Comment (cfezequiel) [inline sklearn/sklearn-template/template/trainer/metadata.py]: Everywhere: I talked more about 2 vs. 4 space indents in another PR comment, but gist is that for continuation lines, 4-space indent is the recommendation. We can discuss either doing 2-default/4-continue or 4/4-default/continue indent convention.

File: sklearn/sklearn-template/template/trainer/metadata.py [hunk 3]
```diff
@@ -19,17 +19,20 @@
...
+  'pclass',
 ]
 
 LABEL = 'survived'
 
 DATASET_SPLIT_KEY = 'dataset_type'
 
+METRIC_FILE_NAME_PREFIX = 'metric'
+MODEL_FILE_NAME_PREFIX = 'model'
+DUMP_FILE_NAME_SUFFIX = '.joblib'
```
Comment (cfezequiel) [inline sklearn/sklearn-template/template/trainer/metadata.py]: Consider renaming to `MODEL_FILE_NAME_SUFFIX`

File: sklearn/sklearn-template/template/trainer/task.py [hunk 4]
```diff
@@ -15,28 +15,85 @@
...
+  """
+  storage_client = storage.Client()
+  parse_result = urlparse(gcs_path)
 
-from trainer import input_util
-from trainer import model
+  # Parse bucket name
+  gcs_path = parse_result.path
+  bucket_name = parse_result.hostname
+  bucket = storage_client.get_bucket(bucket_name)
```
Comment (cfezequiel) [inline sklearn/sklearn-template/template/trainer/task.py:57]: I'd recommend the `tensorflow.gfile` approach as it's simpler (see my comments in PR #4 ), unless there is an advantage you see in using `storage`.

File: sklearn/sklearn-template/template/trainer/task.py [hunk 5]
```diff
@@ -46,18 +103,41 @@ def _train_and_evaluate(estimator, dataset, output_dir):
...
+  hpt.report_hyperparameter_tuning_metric(
+    hyperparameter_metric_tag='my_metric_tag',
+    metric_value=np.mean(scores),
+    global_step=1000)
+
+  timestamp = str(int(time.time()))
+  trial_id = str(hpt.trial_id)
+
+  # Export to the folder of output_dir/trial_id/FILE_NAME_PREFIX_timestampFILE_NAME_SUFFIX
+  model_output_path = os.path.join(output_dir, trial_id,
```
Comment (cfezequiel) [inline sklearn/sklearn-template/template/trainer/task.py]: Suggest using `'{}_{}{}'.format(...)` to form the strings.

File: sklearn/sklearn-template/template/trainer/task.py [hunk 6]
```diff
@@ -70,21 +150,59 @@ def _parse_args(argv):
...
+      'WARN'
+    ],
+    default='INFO',
+  )
+
+  parser.add_argument(
+    '--n_estimator',
+    help='Number of trees in the forest.',
+    default=10,
+    type=int
```
Comment (cfezequiel) [inline sklearn/sklearn-template/template/trainer/task.py]: Everywhere: add comma to last item of a list when using multi-line form, e.g. `type=int,`

File: sklearn/sklearn-template/template/trainer/task.py (hunk 4 above)
Comment (luotigerlsx) [reply sklearn/sklearn-template/template/trainer/task.py:57]: There are two reasons that I think we shouldn't use tf.gfile
1. All the other examples in sklearn/ folder don't use tf.gfile, I think we should align to them
2. If people choose to use sklearn, in a lot of cases I would assume that they don't have tensorflow experience or knowledge. Even just making use of tf.gfile may still confuse people.

Therefore, I remove the tensorflow dependency completely. What do you think?

File: sklearn/sklearn-template/template/setup.py (hunk 1 above)
Comment (luotigerlsx) [reply sklearn/sklearn-template/template/setup.py]: Modified as recommended.

File: sklearn/sklearn-template/template/trainer/metadata.py (hunk 2 above)
Comment (luotigerlsx) [reply sklearn/sklearn-template/template/trainer/metadata.py]: Modified as 2-default/4-continue. Will change other model and input_util in the other PR to avoid conflict.

File: sklearn/sklearn-template/template/trainer/metadata.py (hunk 3 above)
Comment (luotigerlsx) [reply sklearn/sklearn-template/template/trainer/metadata.py]: Modified as recommended.

File: sklearn/sklearn-template/template/trainer/task.py (hunk 5 above)
Comment (luotigerlsx) [reply sklearn/sklearn-template/template/trainer/task.py]: Modified as recommended.

File: sklearn/sklearn-template/template/trainer/task.py (hunk 6 above)
Comment (luotigerlsx) [reply sklearn/sklearn-template/template/trainer/task.py]: Modified as recommended.

File: sklearn/sklearn-template/template/setup.py [hunk 7]
```diff
@@ -17,18 +17,16 @@
...
-
 REQUIRED_PACKAGES = [
     'matplotlib>=2.2.3',
     'seaborn>=0.9.0',
     'scikit-learn>=0.20.2',
     'pandas-gbq>=0.8.0',
     'cloudml-hypertune',
     'scikit-plot',
-    "urllib"
+    "urllib",
```
Comment (cfezequiel) [inline sklearn/sklearn-template/template/setup.py:29]: Suggest changing `"` to `'` for consistency with the rest of the code.

File: sklearn/sklearn-template/template/trainer/task.py (hunk 4 above)
Comment (cfezequiel) [reply sklearn/sklearn-template/template/trainer/task.py:57]: Thanks for the feedback. Here are my thoughts regarding your points:
1. I understand. We would be changing the convention then. We could let the repo owners decide if they would want to go with this approach or keep the existing one.
2. Right, actually `gfile` is more closely related to the Python `file` module than anything in `tensorflow`. So they wouldn't really need to know anything about TensorFlow at all, other than the fact that `gfile` exists and they could read more about it in: https://www.tensorflow.org/api_docs/python/tf/gfile

While it's a bit unfortunate that they placed a great file handling module `gfile` into `tensorflow` specifically, I think we should still use it because it reduces the complexity of reading/writing files from/to GCS. It takes about a couple of lines of code to do GCS file I/O with `gfile`.

Practically speaking, if you were a user (say a data scientist), and you learned about `gfile`, would you still want to use `storage` for simple GCS file I/O?

Also, I think it would be nice to get users thinking about `tensorflow` anyway, even though the main framework they're using is `sklearn`. :)

And we could use `from tensorflow import gfile` import convention so we don't need to propagate `tf` throughout the rest of the code.
